### PR TITLE
[Dataset] Contribute IGBH dataset to hetero_rgcn.py

### PR DESCRIPTION
**Pull request by BowenYao18**
## Description
I added IGB-dataset to the RGCN folder. Currently the three smaller version (tiny[~2GB], small[~20GB], medium[~100GB]) are supported. The two larger versions (large[~800GB], full[~2.2TB]) are still under test. 

Usage:
1. Run the "download.py --size {choose the size}", the dataset will be downloaded and processed under the default ""dataset/"" folder
2. Run the ""hetero_rgcn.py"" with the corresponding dataset you just processed

## Checklist
Please feel free to remove inapplicable items for your PR.
- [Y] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [Y] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [Y] Code is well-documented
- [Y] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [Y] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/graphbolt/rgcn/download.py [hunk 1]
```diff
@@ -0,0 +1,608 @@
+import argparse, hashlib, os, shutil, tarfile, yaml
+import subprocess
+import urllib.request as ur
+
+import dgl.graphbolt as gb
+import numpy as np
+from tqdm import tqdm
+
+GBFACTOR = float(1 << 30)
```
Comment (mfbalin) [inline examples/graphbolt/rgcn/download.py:9]:
```suggestion
GBFACTOR = 1 << 30
```

File: examples/graphbolt/rgcn/download.py [hunk 2]
```diff
@@ -0,0 +1,608 @@
...
+    gb.numpy_save_aligned(file_path, array)
+
+    # Assert the shape and elements of the array are correct
+    if num_class == 2983 and dataset_size == "full":
+        # new_array = np.memmap(file_path, dtype='int32', mode='r',  shape=(227130858, 1))
+        new_array = np.load(file_path)
+        assert new_array.shape[0] == 227130858
+        assert np.array_equal(array, new_array)
+    else:
+        # new_array = np.memmap(file_path, dtype='int32', mode='r',  shape=(num_nodes[dataset_size]["paper"], 1))
```
Comment (mfbalin) [inline examples/graphbolt/rgcn/download.py:483]:
```suggestion
        assert num_class == 19
        # new_array = np.memmap(file_path, dtype='int32', mode='r',  shape=(num_nodes[dataset_size]["paper"], 1))
```

File: examples/graphbolt/rgcn/download.py [hunk 3]
```diff
@@ -0,0 +1,608 @@
...
+}
+
+
+def split_data(label_path, set_dir, dataset_size):
+    """This is for splitting the labels into three sets: train, validation, and test sets."""
+    # labels = np.memmap(label_path, dtype='int32', mode='r',  shape=(num_nodes[dataset_size]["paper"], 1))
+    labels = np.load(label_path)
+
+    total_samples = len(labels)
+    train_end = int(0.8 * total_samples)
```
Comment (mfbalin) [inline examples/graphbolt/rgcn/download.py:399]:
```suggestion
    train_end = int(0.6 * total_samples)
```

File: examples/graphbolt/rgcn/download.py [hunk 4]
```diff
@@ -0,0 +1,608 @@
...
+    validation_end = int(0.9 * total_samples)
```
Comment (mfbalin) [inline examples/graphbolt/rgcn/download.py:400]:
```suggestion
    validation_end = int(0.8 * total_samples)
```

File: examples/graphbolt/rgcn/download.py [hunk 5]
```diff
@@ -0,0 +1,608 @@
...
+                },
+                {
+                    "path": "edges/paper__topic__fos.npy",
+                    "type": "paper:has_topic:field_of_study",
+                    "format": "numpy",
+                },
+            ],
+        },
+        "tasks": [
+            {
```
Comment (mfbalin) [inline examples/graphbolt/rgcn/download.py:52]: Can we also add the 2nd 2K task? One can have the name `node_classification_19`, the other can have the name `node_classification_2983`. Or if you have an idea on how to better name these 2 alternatives, you can go ahead.

Comment (BowenYao18) [reply examples/graphbolt/rgcn/download.py:52]: Sure. I will add another task on 2983 classes

File: examples/graphbolt/rgcn/download.py (hunk 2 above)
Comment (BowenYao18) [reply examples/graphbolt/rgcn/download.py:484]: For this one, the else condition is for all cases other cases excluding the case (num_class=2983 && size == "full"), so I should comment out this assert?

Comment (mfbalin) [reply examples/graphbolt/rgcn/download.py:484]: Don't all datasets work with both 2983 and 19?

File: examples/graphbolt/rgcn/download.py (hunk 5 above)
Comment (mfbalin) [reply examples/graphbolt/rgcn/download.py:52]: 1 task 1 run. In the training script, we can choose the task[1] too if we want to.

File: examples/graphbolt/rgcn/download.py [hunk 6]
```diff
@@ -0,0 +1,686 @@
...
+                                "in_memory": in_memory,
+                                "path": "set/validation_labels_19.npy",
+                                "name": "labels",
+                                "format": "numpy",
+                            },
+                        ],
+                        "type": "paper",
+                    }
+                ],
+                "name": "node_classification_19",
```
Comment (mfbalin) [inline examples/graphbolt/rgcn/download.py:73]: If the other is going to be 2K, then maybe we don't need 19 suffix here.
```suggestion
                "name": "node_classification",
```

File: examples/graphbolt/rgcn/download.py (hunk 5 above)
Comment (mfbalin) [reply examples/graphbolt/rgcn/download.py:52]: But it is also technically possible to do 1 run 2 tasks at the same time.

Comment (BowenYao18) [reply examples/graphbolt/rgcn/download.py:52]: Got it. Do we need to do 1 run 2 tasks or it is fine to keep it 1 task 1 run for now. Currently we should support to run the 2983 task asa long as we change tasks[0] to tasks[1]

File: examples/graphbolt/rgcn/download.py [hunk 7]
```diff
@@ -0,0 +1,686 @@
...
+                "in_memory": in_memory,
+                "path": "data/paper_feat.npy",
+                "type": "paper",
+            },
+            {
+                "domain": "node",
+                "name": "label",
+                "format": "numpy",
+                "in_memory": in_memory,
+                "path": "data/paper_label_19.npy",
```
Comment (mfbalin) [inline examples/graphbolt/rgcn/download.py:189]: Why is there no 2K labels here? Also, are labels also added as features to the dataset?

File: examples/graphbolt/rgcn/download.py (hunk 5 above)
Comment (mfbalin) [reply examples/graphbolt/rgcn/download.py:52]: 1 run 1 task is more than sufficient in the example so long as the dataset contains both tasks.

File: examples/graphbolt/rgcn/download.py (hunk 7 above)
Comment (BowenYao18) [reply examples/graphbolt/rgcn/download.py:189]: I will add the 2983 now, and yes, I added it as feature data. I am not sure where to add the label

Comment (mfbalin) [reply examples/graphbolt/rgcn/download.py:189]: The labels are part of the tasks?

Comment (BowenYao18) [reply examples/graphbolt/rgcn/download.py:189]: Yes. But the label file is already processed into the three sets, so the original whole label file is not needed

Comment (mfbalin) [reply examples/graphbolt/rgcn/download.py:189]: @frozenbugs Do we need to add labels as features?

File: examples/graphbolt/rgcn/hetero_rgcn.py [hunk 8]
```diff
@@ -663,7 +670,15 @@ def main(args):
...
         default="ogbn-mag",
-        choices=["ogbn-mag", "ogb-lsc-mag240m"],
+        choices=[
+            "ogbn-mag",
+            "ogb-lsc-mag240m",
+            "igb-heterogeneous-tiny",
+            "igb-heterogeneous-small",
+            "igb-heterogeneous-medium",
+            "igb-heterogeneous-large",
+            "igb-heterogeneous-full",
```
Comment (mfbalin) [inline examples/graphbolt/rgcn/hetero_rgcn.py:680]: Do you think it is better to follow the convention in the [paper](https://dl.acm.org/doi/pdf/10.1145/3580305.3599843)?
```suggestion
            "igb-het-tiny",
            "igb-het-small",
            "igb-het-medium",
            "igb-het-large",
            "igb-het",
```

File: examples/graphbolt/rgcn/download.py [hunk 9]
```diff
@@ -0,0 +1,670 @@
...
+                {
+                    "num": num_nodes[dataset_size]["institute"],
+                    "type": "institution",
+                },
+                {
+                    "num": num_nodes[dataset_size]["fos"],
+                    "type": "field_of_study",
+                },
+            ],
+            "edges": [
```
Comment (mfbalin) [inline examples/graphbolt/rgcn/download.py:28]: Let's modify edges as done in this file to add the reverse direction as well:
https://github.com/mlcommons/training/blob/cdd928d4596c142c15a7d86b2eeadbac718c8da2/graph_neural_network/compress_graph.py#L48-L56

Comment (BowenYao18) [reply examples/graphbolt/rgcn/download.py:28]: so we store another version copy of the reverse edge in the edges dir?

Comment (mfbalin) [reply examples/graphbolt/rgcn/download.py:28]: Yes, see how they did here: https://github.com/mlcommons/training/blob/cdd928d4596c142c15a7d86b2eeadbac718c8da2/graph_neural_network/compress_graph.py#L73-L84

Comment (mfbalin) [reply examples/graphbolt/rgcn/download.py:28]: https://mlcommons.org/2024/06/gnn-for-mlperf-training-v4/ here they say they add reverse edges for better accuracy.

Comment (mfbalin) [reply examples/graphbolt/rgcn/download.py:28]: https://github.com/mlcommons/training/blob/cdd928d4596c142c15a7d86b2eeadbac718c8da2/graph_neural_network/compress_graph.py#L47

File: examples/graphbolt/rgcn/evaluator.py
Comment (mfbalin) [inline examples/graphbolt/rgcn/evaluator.py:1]: It looks like the code here can be simplified to a few lines of code. Can we do that?

File: examples/graphbolt/rgcn/hetero_rgcn.py
Comment (mfbalin) [inline examples/graphbolt/rgcn/hetero_rgcn.py:1]: Can you add the igb datasets to the examples/graphbolt/pyg/hetero example as well?

File: examples/graphbolt/pyg/hetero/node_classification.py [hunk 10]
```diff
@@ -400,7 +413,7 @@ def parse_args():
     return parser.parse_args()
 
 
-def main():
+def main(args):
```
Comment (mfbalin) [inline examples/graphbolt/pyg/hetero/node_classification.py:416]: Why is this change necessary?

File: examples/graphbolt/pyg/hetero/node_classification.py [hunk 11]
```diff
@@ -517,4 +530,4 @@ def main():
 
 if __name__ == "__main__":
     args = parse_args()
-    main()
+    main(args)
```
Comment (mfbalin) [inline examples/graphbolt/pyg/hetero/node_classification.py:533]: Same as above.

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 12]
```diff
@@ -990,6 +990,16 @@ class BuiltinDataset(OnDiskDataset):
...
 
+    **igb-het-[tiny|small|medium]**
+        The igb-hom-[tiny|small|medium] dataset is a heterogeneous citation network,
+        which is designed for developers to train and evaluate GNN models with
+        high fidelity. See more details in `igb-het-[tiny|small|medium]
+        <https://github.com/IllinoisGraphBenchmark/IGB-Datasets>`_.
+
+        .. note::
+            Reverse paper__cites__paper edges are added to the original graph.
+            Node features are stored as float32.
```
Comment (mfbalin) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:1001]: Does adding reverse edges help the accuracy? If not, let's add it only for the MLPerf version of the dataset, named `igb-het-mlperf`. Let's also use float16 dtype for the features of `igb-het-mlperf`. Note that there will still be `igb-het`.

Comment (mfbalin) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:1001]: What about other reverse edges?

File: examples/graphbolt/pyg/hetero/node_classification.py [hunk 13]
```diff
@@ -58,12 +58,20 @@ def create_dataloader(
         datapipe = datapipe.copy_to(device=device)
         need_copy = False
 
+    # if args.dataset == "ogb-lsc-mag240m":
+    #     node_feature_keys = {
+    #         "paper": ["feat"],
+    #         "author": ["feat"],
+    #         "institution": ["feat"],
+    #     }
+    node_feature_keys = {"paper": ["feat"]}
```
Comment (mfbalin) [inline examples/graphbolt/pyg/hetero/node_classification.py:67]:
```suggestion
    node_feature_keys = {"paper": ["feat"], "author": ["feat"]}
```

File: examples/graphbolt/pyg/hetero/node_classification.py [hunk 14]
```diff
@@ -58,12 +58,20 @@ def create_dataloader(
...
     if args.dataset == "ogb-lsc-mag240m":
-        node_feature_keys = {
-            "paper": ["feat"],
-            "author": ["feat"],
-            "institution": ["feat"],
-        }
+        node_feature_keys["author"] = ["feat"]
```
Comment (mfbalin) [inline examples/graphbolt/pyg/hetero/node_classification.py:69]:
```suggestion
```

File: examples/graphbolt/pyg/hetero/node_classification.py [hunk 15]
```diff
@@ -58,12 +58,20 @@ def create_dataloader(
...
+        node_feature_keys["institution"] = ["feat"]
+    if "igb-het" in args.dataset:
+        node_feature_keys["author"] = ["feat"]
```
Comment (mfbalin) [inline examples/graphbolt/pyg/hetero/node_classification.py:72]:
```suggestion
```

File: examples/graphbolt/rgcn/evaluator.py
Comment (mfbalin) [inline examples/graphbolt/rgcn/evaluator.py:1]: What does this file have that the existing examples does not have? We already have functions to measure accuracy.

Comment (mfbalin) [reply examples/graphbolt/rgcn/evaluator.py:1]: @BowenYao18

**Comment by mfbalin** [comment]
Extracted your commits to #7788. Merging the dataset so that I can test it.

File: examples/graphbolt/pyg/hetero/node_classification.py [hunk 16]
```diff
@@ -344,8 +344,9 @@ def parse_args():
             "igb-het-tiny",
             "igb-het-small",
             "igb-het-medium",
+            "igb-het-large",
```
Comment (mfbalin) [inline examples/graphbolt/pyg/hetero/node_classification.py:347]: add igb-het and igb-het-mlperf here.

File: examples/graphbolt/rgcn/evaluator.py
Comment (mfbalin) [inline examples/graphbolt/rgcn/evaluator.py:1]: We don't need this file. The same functionality can be implemented in a few lines.

File: examples/graphbolt/rgcn/hetero_rgcn.py [hunk 17]
```diff
@@ -663,8 +663,18 @@ def main(args):
...
-        help="Dataset name. Possible values: ogbn-mag, ogb-lsc-mag240m",
+        choices=[
+            "ogbn-mag",
+            "ogb-lsc-mag240m",
+            "igb-het-tiny",
+            "igb-het-small",
+            "igb-het-medium",
+            "igb-het-large",
+            "igb-het",
+            "igb-het-MLPerf",
```
Comment (mfbalin) [inline examples/graphbolt/rgcn/hetero_rgcn.py:674]: Can we make it all lower case?

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 18]
```diff
@@ -1047,6 +1047,8 @@ class BuiltinDataset(OnDiskDataset):
         "igb-hom-seeds",
         "igb-het-medium",
         "igb-het-medium-seeds",
+        "igb-het-large",
+        "igb-het-large-seeds",
```
Comment (mfbalin) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:1051]: igb-het and igb-het-mlperf needs to be added here.

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 19]
```diff
@@ -990,10 +990,10 @@ class BuiltinDataset(OnDiskDataset):
             Self edges are added to the original graph.
             Node features are stored as float32.
 
-    **igb-het-[tiny|small|medium]**
+    **igb-het-[tiny|small|medium|large]**
         The igb-hom-[tiny|small|medium] dataset is a heterogeneous citation network,
         which is designed for developers to train and evaluate GNN models with
-        high fidelity. See more details in `igb-het-[tiny|small|medium]
+        high fidelity. See more details in `igb-het-[tiny|small|medium|large]
```
Comment (mfbalin) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:996]: also here as well.

**Comment by mfbalin** [comment]
LGTM except for the overly complicated evaluator.py file.

File: examples/graphbolt/rgcn/evaluator.py
Comment (mfbalin) [inline examples/graphbolt/rgcn/evaluator.py:1]: Let's use pure pytorch operations to compute the evaluation score.

**Review by mfbalin — CHANGES_REQUESTED** [review]
